sqlalchemy-stubs/sql/functions.pyi

```diff
@@ -1,250 +1,308 @@
-    def alias(self, name: Optional[Any] = ...): ...
-    def select(self): ...
-    def scalar(self): ...
-    def execute(self): ...
-    def self_group(self, against: Optional[Any] = ...): ...
-
-class FunctionAsBinary(BinaryExpression):
-    sql_function: Any = ...
-    left_index: Any = ...
-    right_index: Any = ...
+    ) -> Over[_TE]: ...
+    def within_group(self, *order_by: Any) -> WithinGroup[_TE]: ...
+    @overload
+    def filter(self) -> _FE: ...
+    @overload
+    def filter(
+        self, criteria: Any, *criterion: Any
+    ) -> FunctionFilter[_TE]: ...
+    def as_comparison(
+        self, left_index: int, right_index: int
+    ) -> FunctionAsBinary[_TE]: ...
+    def within_group_type(
+        self, within_group: Any
+    ) -> Optional[type_api.TypeEngine[Any]]: ...
+    def alias(self, name: Optional[Any] = ...) -> TableValuedAlias: ...  # type: ignore[override]
+    def select(self) -> Select: ...  # type: ignore[override]
+    def scalar(self) -> Any: ...  # type: ignore[override]
+    def execute(self) -> Any: ...  # type: ignore[override]
+    def self_group(
+        self: _FE, against: Optional[Any] = ...
+    ) -> Union[_FE, Grouping[_TE], AsBoolean[_FE]]: ...
+
+class FunctionAsBinary(BinaryExpression[_TE]):
+    sql_function: FunctionElement[_TE] = ...
+    left_index: int = ...
+    right_index: int = ...
     operator: Any = ...
     type: Any = ...
     negate: Any = ...
     modifiers: Any = ...
-    def __init__(self, fn: Any, left_index: Any, right_index: Any) -> None: ...
-    @property
-    def left(self): ...
-    @left.setter
-    def left(self, value: Any) -> None: ...
-    @property
-    def right(self): ...
-    @right.setter
-    def right(self, value: Any) -> None: ...
+    left: ClauseElement = ...
+    right: ClauseElement = ...
+    def __init__(
+        self, fn: FunctionElement[_TE], left_index: int, right_index: int
+    ) -> None: ...
 
-class ScalarFunctionColumn(NamedColumn):
+class ScalarFunctionColumn(NamedColumn[_TE]):
     __visit_name__: str = ...
     is_literal: bool = ...
     table: Any = ...
-    fn: Any = ...
+    fn: ClauseElement = ...
     name: Any = ...
-    type: Any = ...
+    type: _TE = ...  # type: ignore[assignment]
+    @overload
     def __init__(
-        self, fn: Any, name: Any, type_: Optional[Any] = ...
+        self: ScalarFunctionColumn[sqltypes.NullType],
+        fn: ClauseElement,
+        name: str,
+        type_: None = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self, fn: ClauseElement, name: str, type_: Union[_TE, Type[_TE]]
     ) -> None: ...
 
 class _FunctionGenerator:
     opts: Any = ...
     def __init__(self, **opts: Any) -> None: ...
-    def __getattr__(self, name: Any): ...
-    def __call__(self, *c: Any, **kwargs: Any): ...
-
-func: _FunctionGenerator = ...
-modifier: _FunctionGenerator = ...
-
-class Function(FunctionElement):
+    def __getattr__(self, name: Any) -> Any: ...
```

Comment:
this is assumed to be `name: str` also I think this could return `_FunctionGenerator`